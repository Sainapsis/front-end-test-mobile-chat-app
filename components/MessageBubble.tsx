import React from 'react';
import { View, StyleSheet } from 'react-native';
import { ThemedText } from './ThemedText';
import { Message } from '@/hooks/useChats';
import { useColorScheme } from '@/hooks/useColorScheme';

interface MessageBubbleProps {
  message: Message;
  isCurrentUser: boolean;
}

export function MessageBubble({ message, isCurrentUser }: MessageBubbleProps) {
  const colorScheme = useColorScheme();
  const isDark = colorScheme === 'dark';

  const formatTime = (timestamp: number) => {
    const date = new Date(timestamp);
    return date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
  };

  return (
    <View style={[
      styles.container,
      isCurrentUser ? styles.selfContainer : styles.otherContainer
    ]}
    >
      <View style={[
        styles.bubble,
        isCurrentUser 
          ? [styles.selfBubble, { backgroundColor: isDark ? '#235A4A' : '#DCF8C6' }]
          : [styles.otherBubble, { backgroundColor: isDark ? '#2A2C33' : '#FFFFFF' }]
      ]}>
        <ThemedText style={[
          styles.messageText,
          isCurrentUser && !isDark && styles.selfMessageText
        ]}>
          {message.text}
        </ThemedText>
        <View style={styles.timeContainer}>
          <ThemedText style={styles.timeText}>
            {formatTime(message.timestamp)}
          </ThemedText>
        </View>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    marginVertical: 4,
    maxWidth: '80%',
  },
  selfContainer: {
    alignSelf: 'flex-end',
  },
  otherContainer: {
    alignSelf: 'flex-start',
  },
  bubble: {
    borderRadius: 16,
    paddingHorizontal: 12,
    paddingVertical: 8,
    elevation: 1,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.1,
    shadowRadius: 1,
  },
  selfBubble: {
    borderBottomRightRadius: 4,
  },
  otherBubble: {
    borderBottomLeftRadius: 4,
  },
  messageText: {
    fontSize: 16,
  },
  selfMessageText: {
    color: '#000000',
  },
  timeContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    marginTop: 2,
  },
  timeText: {
    fontSize: 11,
    opacity: 0.7,
  },
}); 