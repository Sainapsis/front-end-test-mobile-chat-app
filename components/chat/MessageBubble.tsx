import React, { useState, useEffect, useRef } from 'react';
import { View, StyleSheet, Image, Pressable, Modal, Alert, Animated } from 'react-native';
import { ThemedText } from '@/components/common/ThemedText';
import { Message } from '@/interfaces/chatTypes';
import { useColorScheme } from '@/hooks/useColorScheme';
import { MessageStatus } from './MessageStatus';
import { IconSymbol } from '@/components/ui/IconSymbol';
import { ImageViewer } from './ImageViewer';
import { useAppContext } from '@/hooks/AppContext';
import { useVideoPlayer, VideoView } from 'expo-video';

interface MessageBubbleProps {
  message: Message;
  isCurrentUser: boolean;
  onDelete?: (messageId: string) => void;
}

export function MessageBubble({ message, isCurrentUser, onDelete }: MessageBubbleProps) {
  const isDark = useColorScheme() === 'dark';
  const [isImageViewerVisible, setIsImageViewerVisible] = useState(false);
  const [isVideoViewerVisible, setIsVideoViewerVisible] = useState(false);
  const { theme } = useAppContext();
  const fadeAnim = useRef(new Animated.Value(1)).current;

  const videoPlayer = useVideoPlayer(
    message.mediaType === 'video' && message.mediaUrl ? message.mediaUrl : 'https://example.com/placeholder.mp4',
    player => {
      player.loop = false;
    }
  );

  const formatTime = (timestamp: number) => {
    return new Date(timestamp).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
  };

  const formatFileSize = (bytes?: number | null) => {
    if (!bytes) return '';
    const sizes = ['B', 'KB', 'MB', 'GB'];
    const i = Math.floor(Math.log(bytes) / Math.log(1024));
    return `${(bytes / Math.pow(1024, i)).toFixed(1)} ${sizes[i]}`;
  };

  const renderMedia = () => {
    if (!message.mediaType || !message.mediaUrl) return null;

    switch (message.mediaType) {
      case 'image':
        return (
          <Pressable onPress={() => setIsImageViewerVisible(true)}>
            <Image
              source={{ uri: message.mediaUrl }}
              style={styles.media}
              resizeMode="cover"
            />
          </Pressable>
        );
      case 'video':
        return (
          <Pressable onPress={() => setIsVideoViewerVisible(true)}>
            <View style={styles.videoContainer}>
              <View style={styles.videoThumbnail}>
                <IconSymbol name="videocam" size={32} color={theme.colors.tint} />
                <View style={styles.videoInfo}>
                  <ThemedText style={styles.videoText}>
                    {message.mediaName || 'Video'}
                  </ThemedText>
                  {message.mediaSize && (
                    <ThemedText style={styles.videoSize}>
                      {formatFileSize(message.mediaSize)}
                    </ThemedText>
                  )}
                </View>
              </View>
            </View>
          </Pressable>
        );
      case 'audio':
        return (
          <View style={styles.audioContainer}>
            <IconSymbol name="musical-note" size={24} color={theme.colors.tint} />
            <ThemedText style={styles.audioText}>
              {message.mediaName || 'Audio'}
            </ThemedText>
          </View>
        );
      case 'file':
        return (
          <View style={styles.fileContainer}>
            <IconSymbol name="document" size={24} color={theme.colors.tint} />
            <ThemedText style={styles.fileText}>
              {message.mediaName || 'Archivo'}
            </ThemedText>
          </View>
        );
      default:
        return null;
    }
  };

  const bubbleStyle = [
    styles.bubble,
    isCurrentUser
      ? [styles.selfBubble, { backgroundColor: isDark ? '#E2E2B6' : '#E2E2B6' }]
      : [styles.otherBubble, { backgroundColor: isDark ? '#2A2C33' : '#ffff' }]
  ];

  const handleLongPress = () => {
    if (isCurrentUser) {
      Alert.alert(
        'Eliminar mensaje',
        '¿Estás seguro de que quieres eliminar este mensaje?',
        [
          {
            text: 'Cancelar',
            style: 'cancel',
          },
          {
            text: 'Eliminar',
            style: 'destructive',
            onPress: () => {
              Animated.timing(fadeAnim, {
                toValue: 0,
                duration: 300,
                useNativeDriver: true,
              }).start(() => {
                onDelete?.(message.id);
              });
            },
          },
        ]
      );
    }
  };

  useEffect(() => {
    if (!isVideoViewerVisible) {
      videoPlayer.pause();
    }
  }, [isVideoViewerVisible]);

  return (
    <>
      <Animated.View style={{ opacity: fadeAnim }}>
        <Pressable
          onLongPress={handleLongPress}
          delayLongPress={500}
          style={[styles.container, isCurrentUser ? styles.selfContainer : styles.otherContainer]}
        >
          <View style={bubbleStyle}>
            {renderMedia()}
            {message.text && (
              <ThemedText style={[styles.messageText, isCurrentUser && { color: '#000000' }]}>
                {message.text}
              </ThemedText>
            )}
            <View style={styles.footer}>
              <ThemedText style={[styles.timeText, isCurrentUser && { color: '#000000' }]}>
                {formatTime(message.timestamp)}
              </ThemedText>
              {isCurrentUser && (
                <View style={styles.footerActions}>
                  <MessageStatus message={message} />
                </View>
              )}
            </View>
          </View>
        </Pressable>
      </Animated.View>

      <ImageViewer
        visible={isImageViewerVisible}
        onClose={() => setIsImageViewerVisible(false)}
        imageUri={message.mediaUrl || ''}
      />

      <Modal
        visible={isVideoViewerVisible}
        transparent={true}
        animationType="fade"
        onRequestClose={() => setIsVideoViewerVisible(false)}
      >
        <View style={styles.modalContainer}>
          <Pressable
            style={styles.closeButton}
            onPress={() => setIsVideoViewerVisible(false)}
          >
            <IconSymbol name="close" size={24} color="#FFFFFF" />
          </Pressable>
          <VideoView
            player={videoPlayer}
            style={styles.fullScreenVideo}
            allowsFullscreen
            allowsPictureInPicture
          />
        </View>
      </Modal>
    </>
  );
}

const styles = StyleSheet.create({
  container: {
    marginVertical: 4,
    maxWidth: '80%',
  },
  selfContainer: {
    alignSelf: 'flex-end',
  },
  otherContainer: {
    alignSelf: 'flex-start',
  },
  bubble: {
    borderRadius: 16,
    padding: 12,
    elevation: 1,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.1,
    shadowRadius: 1,
  },
  selfBubble: {
    borderBottomRightRadius: 4,
  },
  otherBubble: {
    borderBottomLeftRadius: 4,
  },
  messageText: {
    fontSize: 16,
  },
  selfMessageText: {
    color: '#000000',
  },
  footer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
    marginTop: 2,
  },
  timeText: {
    fontSize: 11,
    opacity: 0.7,
  },
  media: {
    width: 250,
    height: 200,
    borderRadius: 12,
  },
  videoContainer: {
    width: 250,
    height: 200,
    borderRadius: 12,
    overflow: 'hidden',
    backgroundColor: '#F0F0F0',
  },
  videoThumbnail: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: 'rgba(0,0,0,0.1)',
  },
  videoInfo: {
    position: 'absolute',
    bottom: 0,
    left: 0,
    right: 0,
    padding: 8,
    backgroundColor: 'rgba(0,0,0,0.7)',
  },
  videoText: {
    fontSize: 14,
    color: '#FFFFFF',
  },
  videoSize: {
    fontSize: 12,
    color: '#FFFFFF',
    opacity: 0.7,
  },
  audioContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 8,
    backgroundColor: '#000000',
    borderRadius: 8,
    marginTop: 4,
    gap: 8,
  },
  audioText: {
    flex: 1,
    fontSize: 14,
  },
  fileContainer: {
    width: 250,
    borderRadius: 12,
    overflow: 'hidden',
    backgroundColor: '#F0F0F0',
    flexDirection: 'row',
    alignItems: 'center',
    padding: 8,
    marginBottom: 8,
  },
  fileText: {
    flex: 1,
    marginLeft: 8,
    fontSize: 14,
  },
  modalContainer: {
    flex: 1,
    backgroundColor: '#000000',
    justifyContent: 'center',
  },
  closeButton: {
    position: 'absolute',
    top: 40,
    right: 20,
    zIndex: 1,
    padding: 8,
  },
  fullScreenVideo: {
    width: '100%',
    height: '100%',
  },
  footerActions: {
    flexDirection: 'row',
    alignItems: 'center',
    gap: 4,
  },
}); 