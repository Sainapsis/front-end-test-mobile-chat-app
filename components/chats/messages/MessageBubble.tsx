import React from 'react';
import { View, StyleSheet } from 'react-native';
import { ThemedText } from '@/components/ui/text/ThemedText';
import { Message } from '@/hooks/chats/useChats';
import { useColorScheme } from '@/hooks/themes/useColorScheme';
import { Avatar } from '@/components/ui/user/Avatar';
import { User } from '@/hooks/user/useUser';
import { IconSymbol } from '@/components/ui/icons/IconSymbol';

interface MessageBubbleProps {
  message: Message;
  isCurrentUser: boolean;
  isReaded: boolean;
}

export function MessageBubble({ message, isCurrentUser, isReaded }: MessageBubbleProps) {
  const colorScheme = useColorScheme();
  const isDark = colorScheme === 'dark';

  const formatTime = (timestamp: number) => {
    const date = new Date(timestamp);
    return date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
  };

  return (
    <View style={styles.chatContainer}>
      <View style={styles.timeContainer}>
        <ThemedText style={[styles.timeText, isCurrentUser
          ? [styles.selfContainer]
          : [styles.otherContainer, styles.otherText]]}>
          {isCurrentUser ? '' : message.senderName + ' '}{formatTime(message.timestamp)}
          {message.readed ?
            <IconSymbol name="checkmark" color="#808080" size={16}></IconSymbol> : <></>
          }
        </ThemedText>
      </View>
      <View style={[
        styles.container,
        isCurrentUser ? styles.selfContainer : styles.otherContainer
      ]}>
        <View style={styles.bubbleContainer}>
          <View style={styles.avatar}>
            {isCurrentUser ? <></> : <Avatar userName={message.senderName} size={30} status={"online"}></Avatar>}
          </View>
          <View style={[
            styles.bubble,
            isCurrentUser
              ? [styles.selfBubble, { backgroundColor: isDark ? '#1E1E1E' : '#EEF6FF' }]
              : [styles.otherBubble, { backgroundColor: isDark ? '#2A2A2A' : '#FFFFFF' }]
          ]}>
            <ThemedText style={[
              styles.messageText,
              isCurrentUser && !isDark && styles.selfMessageText
            ]}>
              {message.text}
            </ThemedText>
          </View>
        </View>
      </View>
    </View>
  );
}

const styles = StyleSheet.create({
  chatContainer: {
    marginBottom: 10,
  },
  container: {
    marginVertical: 4,
    maxWidth: '80%',
  },
  selfContainer: {
    alignSelf: 'flex-end',
  },
  otherContainer: {
    alignSelf: 'flex-start',
  },
  otherText: {
    marginLeft: 45,
  },
  bubble: {
    borderRadius: 10,
    paddingHorizontal: 12,
    paddingVertical: 8,
    elevation: 1,
    shadowOffset: { width: 1, height: 1 },
    shadowOpacity: 0.1,
    shadowRadius: 1
  },
  bubbleContainer: {
    flexDirection: 'row',
    alignItems: 'flex-end',
  },
  selfBubble: {
  },
  otherBubble: {
  },
  messageText: {
    fontSize: 16,
  },
  selfMessageText: {
    color: '#000000',
  },
  timeContainer: {
    flexDirection: 'column',
    marginTop: 2,
  },
  timeText: {
    fontSize: 11,
    opacity: 0.7,
    lineHeight: 11,
    alignItems: 'center',
  },
  avatar: {
    marginRight: 5
  }
}); 