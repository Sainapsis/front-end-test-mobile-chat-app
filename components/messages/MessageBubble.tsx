import React from 'react';
import { View, StyleSheet, Pressable, Alert } from 'react-native';
import { ThemedText } from '../ThemedText';
import { Message } from '@/hooks/useChatRoomMessage';
import { useColorScheme } from '@/hooks/useColorScheme';
import { useMessageStatus } from '@/hooks/useMessageStatus';
import { MessageStatusIcon } from './MessageStatus';
import { MessageContent } from './MessageContent';

interface MessageBubbleProps {
  message: Message;
  isCurrentUser: boolean;
  onEditMessage?: (message: Message) => void;
  onDeleteMessage?: (messageId: string) => void;
}

export function MessageBubble({ message, isCurrentUser, onEditMessage, onDeleteMessage }: MessageBubbleProps) {
  const colorScheme = useColorScheme();
  const isDark = colorScheme === 'dark';
  const { getStatusIcon, shouldShowStatus } = useMessageStatus(message, isCurrentUser);

  const formatTime = (timestamp: number) => {
    const date = new Date(timestamp);
    return date.toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
  };

  const handleLongPress = () => {
    if (!isCurrentUser) return;
    if (message.isDeleted) return;

    Alert.alert(
      'Options Message',
      '¿What do you want to do with this message?',
      [
        {
          text: 'Edit',
          onPress: () => onEditMessage?.(message),
        },
        {
          text: 'Delete',
          onPress: () => {
            Alert.alert(
              'Delete message',
              '¿Are you sure you want to delete this message??',
              [
                {
                  text: 'Cancel',
                  style: 'cancel',
                },
                {
                  text: 'Delete',
                  style: 'destructive',
                  onPress: () => onDeleteMessage?.(message.id),
                },
              ]
            );
          },
          style: 'destructive',
        },
        {
          text: 'Cancel',
          style: 'cancel',
        },
      ]
    );
  };

  return (
    <Pressable
      onLongPress={handleLongPress}
      style={[
        styles.container,
        isCurrentUser ? styles.selfContainer : styles.otherContainer
      ]}
    >
      <View style={[
        styles.bubble,
        isCurrentUser 
          ? [styles.selfBubble, { backgroundColor: isDark ? '#235A4A' : '#DCF8C6' }]
          : [styles.otherBubble, { backgroundColor: isDark ? '#2A2C33' : '#FFFFFF' }]
      ]}>
        <MessageContent 
          message={message} 
          isDark={isDark} 
          isCurrentUser={isCurrentUser} 
        />
        <View style={styles.timeContainer}>
          <ThemedText style={styles.timeText}>
            {formatTime(message.timestamp)}
          </ThemedText>
          {shouldShowStatus 
          && getStatusIcon() 
          && (
            <MessageStatusIcon 
              status={message.status}
              isDark={isDark}
            />
          )}
        </View>
      </View>
    </Pressable>
  );
}

const styles = StyleSheet.create({
  container: {
    marginVertical: 4,
    maxWidth: '80%',
  },
  selfContainer: {
    alignSelf: 'flex-end',
  },
  otherContainer: {
    alignSelf: 'flex-start',
  },
  bubble: {
    borderRadius: 16,
    paddingHorizontal: 12,
    paddingVertical: 8,
    elevation: 1,
    shadowOffset: { width: 0, height: 1 },
    shadowOpacity: 0.1,
    shadowRadius: 1,
  },
  selfBubble: {
    borderBottomRightRadius: 4,
  },
  otherBubble: {
    borderBottomLeftRadius: 4,
  },
  messageText: {
    fontSize: 16,
  },
  selfMessageText: {
    color: '#000000',
  },
  timeContainer: {
    flexDirection: 'row',
    justifyContent: 'flex-end',
    alignItems: 'center',
    marginTop: 2,
    gap: 4,
  },
  timeText: {
    fontSize: 11,
    opacity: 0.7,
  },
  editedIndicator: {
    fontSize: 11,
    opacity: 0.7,
    fontStyle: 'italic',
    marginTop: 2,
  },
  deletedMessage: {
    fontStyle: 'italic',
    opacity: 0.7,
  },
}); 