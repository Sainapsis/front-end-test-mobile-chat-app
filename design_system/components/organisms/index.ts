export { Avatar } from './Avatar';
export { ChatListItem } from './ChatListItem';
export { ParallaxScrollView } from './ParallaxScrollView';
export { MessageBubble } from './MessageBubble';
export { OptionsMenu } from './OptionsMenu';
export { SearchResults } from './SearchResults';
export { UserList } from './UserList';
export { ChatsList } from './ChatsList';
export { NewChatModal } from './NewChatModal';
