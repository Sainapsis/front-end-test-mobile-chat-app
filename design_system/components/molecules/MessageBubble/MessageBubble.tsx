import React from 'react';
import { View, Alert, TouchableWithoutFeedback } from 'react-native';
import { ThemedText } from '@/design_system/components/atoms';
import { Message } from '@/hooks/useChats';
import { useColorScheme } from '@/hooks/useColorScheme';
import { styles, getBubbleColors } from './MessageBubble.styles';

interface MessageBubbleProps {
  message: Message;
  isCurrentUser: boolean;
  onDeleteMessage?: (messageId: string) => void;
}

export function MessageBubble({ message, isCurrentUser, onDeleteMessage }: MessageBubbleProps) {
  const colorScheme = useColorScheme();
  const isDark = colorScheme === 'dark';
  const bubbleColors = getBubbleColors(isDark, isCurrentUser);

  const handleLongPress = () => {
    if (isCurrentUser && onDeleteMessage) {
      Alert.alert(
        'Delete Message',
        'Do you want to delete this message?',
        [
          { text: 'Cancel', style: 'cancel' },
          { text: 'Delete', onPress: () => onDeleteMessage(message.id), style: 'destructive' }
        ]
      );
    }
  };

  const formatTime = (timestamp: number) => {
    return new Date(timestamp).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
  };

  return (
    <TouchableWithoutFeedback onLongPress={handleLongPress}>
      <View style={[styles.container, isCurrentUser ? styles.selfContainer : styles.otherContainer]}>
        <View style={[
          styles.bubble,
          isCurrentUser ? styles.selfBubble : styles.otherBubble,
          { backgroundColor: bubbleColors.background }
        ]}>
          <ThemedText style={[styles.messageText, isCurrentUser && !isDark && styles.selfMessageText]}>
            {message.text}
          </ThemedText>
          <View style={styles.timeContainer}>
            <ThemedText style={styles.timeText}>
              {formatTime(message.timestamp)}
            </ThemedText>
          </View>
        </View>
      </View>
    </TouchableWithoutFeedback>
  );
}
