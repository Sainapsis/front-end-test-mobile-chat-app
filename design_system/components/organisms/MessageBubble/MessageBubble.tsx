import React, { useRef, useState } from 'react';
import { View, TouchableWithoutFeedback, Modal } from 'react-native';
import { ThemedText } from '@/design_system/components/atoms';
import { styles } from './MessageBubble.styles';
import { useMessageBubble } from '@/hooks/components/useMessageBubble';
import { Message } from '@/hooks/useChats';
import EmojiSelector, { Categories } from 'react-native-emoji-selector';
import { OptionsMenu } from '@/design_system/components/organisms';

interface MessageBubbleProps {
  message: Message;
  isCurrentUser: boolean;
  userId: string;
  onDeleteMessage?: (messageId: string) => void;
  onAddReaction?: (messageId: string, emoji: string) => void;
  onRemoveReaction?: (reactionId: string, messageId: string) => void;
  onEditMessage?: (messageId: string, currentText: string) => void;
}

export function MessageBubble({ 
  message, 
  isCurrentUser,
  userId,
  onDeleteMessage,
  onAddReaction,
  onRemoveReaction,
  onEditMessage,
}: MessageBubbleProps) {
  const { 
    isDark,
    bubbleColors,
    handleLongPress,
    showEmojiSelector,
    setShowEmojiSelector,
    handleEmojiSelected,
    handleRemoveReaction,
    showOptionsMenu,
    setShowOptionsMenu
  } = useMessageBubble({ 
    message, 
    isCurrentUser,
    userId,
    onDeleteMessage,
    onAddReaction,
    onEditMessage,
    onRemoveReaction 
  });

  const bubbleRef = useRef<View>(null);
  const [bubblePosition, setBubblePosition] = useState({ x: 0, y: 0, width: 0 });

  const formatTime = (timestamp: number) => {
    return new Date(timestamp).toLocaleTimeString([], { hour: '2-digit', minute: '2-digit' });
  };

  const handleLayout = () => {
    bubbleRef.current?.measure((x, y, width, height, pageX, pageY) => {
      setBubblePosition({ x: pageX, y: pageY, width });
    });
  };

  return (
    <>
      <TouchableWithoutFeedback onLongPress={handleLongPress}>
        <View 
          ref={bubbleRef} 
          onLayout={handleLayout} 
          style={[styles.container, isCurrentUser ? styles.selfContainer : styles.otherContainer]}
        >
          <View style={[
            styles.bubble,
            isCurrentUser ? styles.selfBubble : styles.otherBubble,
            { backgroundColor: bubbleColors.background }
          ]}>
            <ThemedText style={[styles.messageText, isCurrentUser && !isDark && styles.selfMessageText]}>
              {message.text}
            </ThemedText>
            <View style={styles.timeContainer}>
              <ThemedText style={styles.timeText}>
                {formatTime(message.timestamp)}
              </ThemedText>
            </View>
            {message.reactions && message.reactions.length > 0 && (
              <View style={[
                styles.reactionsContainer,
                isCurrentUser ? styles.reactionsRight : styles.reactionsLeft
              ]}>
                {message.reactions.map((reaction) => (
                  <TouchableWithoutFeedback
                    key={reaction.id}
                    onPress={() => handleRemoveReaction(reaction.id)}
                  >
                    <View style={styles.reaction}>
                      <ThemedText style={styles.reactionText}>{reaction.emoji}</ThemedText>
                    </View>
                  </TouchableWithoutFeedback>
                ))}
              </View>
            )}
          </View>
        </View>
      </TouchableWithoutFeedback>

      <OptionsMenu
        visible={showOptionsMenu}
        onClose={() => setShowOptionsMenu(false)}
        onEdit={() => onEditMessage?.(message.id, message.text)}
        onDelete={() => onDeleteMessage?.(message.id)}
        onAddEmoji={() => setShowEmojiSelector(true)}
        position={{ top: bubblePosition.y, left: bubblePosition.x, width: bubblePosition.width }} // Pass position as a prop
      />

      <Modal
        visible={showEmojiSelector}
        transparent={true}
        animationType="slide"
        onRequestClose={() => setShowEmojiSelector(false)}
      >
        <TouchableWithoutFeedback onPress={() => setShowEmojiSelector(false)}>
          <View style={styles.modalOverlay}>
            <View style={styles.emojiSelectorContainer}>
              <View style={styles.emojiSelectorHeader}>
                <ThemedText darkColor='#000000'>Select Reaction</ThemedText>
              </View>
              <View style={{ height: 300 }}>
                <EmojiSelector
                  onEmojiSelected={handleEmojiSelected}
                  showSearchBar={false}
                  showHistory={false}
                  showTabs
                  columns={8}
                  category={Categories.emotion}
                  showSectionTitles={false}
                />
              </View>
            </View>
          </View>
        </TouchableWithoutFeedback>
      </Modal>
    </>
  );
}
