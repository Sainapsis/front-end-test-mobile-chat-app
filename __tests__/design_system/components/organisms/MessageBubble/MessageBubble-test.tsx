import * as React from 'react';
import { render, fireEvent } from '@testing-library/react-native';
import renderer from 'react-test-renderer';
import { ThemeProvider } from '@/context/ThemeContext';
import { MessageBubble } from '@/design_system/components/organisms/MessageBubble';
import { Message } from '@/types/Chat';

jest.mock('@/design_system/components/atoms/ThemedText', () => {
    const React = require('react');
    const { Text } = require('react-native');
    return {
      ThemedText: ({ children }: { children: React.ReactNode }) => <Text>{children}</Text>,
    };
  });
  
  jest.mock('@/design_system/components/organisms/OptionsMenu', () => {
    const React = require('react');
    const { View } = require('react-native');
    return {
      OptionsMenu: ({ visible, onClose, onEdit, onDelete, onAddEmoji }: 
        { visible: boolean; onClose: () => void; onEdit: () => void; onDelete: () => void; onAddEmoji: () => void }) => 
        <View testID="options-menu" />,
    };
  });
  
  jest.mock('react-native-emoji-selector', () => {
    const React = require('react');
    const { View } = require('react-native');
    return {
      __esModule: true,
      default: ({ onEmojiSelected }: { onEmojiSelected: (emoji: string) => void }) => 
        <View testID="emoji-selector" />,
      Categories: { emotion: 'emotion' },
    };
  });
  
  jest.mock('@/hooks/components/useMessageBubble', () => ({
    useMessageBubble: jest.fn(() => ({
      isDark: false,
      bubbleColors: { background: '#E5E5EA' },
      handleLongPress: jest.fn(),
      showEmojiSelector: false,
      setShowEmojiSelector: jest.fn(),
      handleEmojiSelected: jest.fn(),
      handleRemoveReaction: jest.fn(),
      showOptionsMenu: false,
      setShowOptionsMenu: jest.fn(),
    })),
  }));
  

describe('MessageBubble Component', () => {
  const mockMessage: Message = {
    id: '123',
    text: 'Hello, this is a test message',
    timestamp: 1620000000000,
    senderId: 'user1',
    reactions: [{ id: 'r1', emoji: 'üëç', userId: 'user2', createdAt: Date.now() }],
  };

  const defaultProps = {
    message: mockMessage,
    isCurrentUser: true,
    userId: 'user1',
    onDeleteMessage: jest.fn(),
    onAddReaction: jest.fn(),
    onRemoveReaction: jest.fn(),
    onEditMessage: jest.fn(),
  };

  const renderWithTheme = (component: React.ReactNode) =>
    render(<ThemeProvider>{component}</ThemeProvider>);

  it('renders correctly with snapshot', () => {
    const tree = renderer.create(
      <ThemeProvider>
        <MessageBubble {...defaultProps} />
      </ThemeProvider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('displays message text correctly', () => {
    const { getByText } = renderWithTheme(<MessageBubble {...defaultProps} />);
    getByText('Hello, this is a test message');
  });

  it('formats and displays timestamp correctly', () => {
    const mockToLocaleTimeString = jest
      .spyOn(Date.prototype, 'toLocaleTimeString')
      .mockReturnValue('10:00 AM');

    const { getByText } = renderWithTheme(<MessageBubble {...defaultProps} />);
    getByText('10:00 AM');

    mockToLocaleTimeString.mockRestore();
  });

  it('renders reactions when present', () => {
    const { getByText } = renderWithTheme(<MessageBubble {...defaultProps} />);
    getByText('üëç');
  });

  it('applies different styles for current user vs other users', () => {
    const { rerender, getByTestId } = renderWithTheme(
      <MessageBubble {...defaultProps} isCurrentUser={true} />
    );
  
    const selfMessageContainer = getByTestId('message-bubble-container');
    expect(selfMessageContainer).toBeTruthy(); // Asegura que existe
  
    const selfStyle = selfMessageContainer.props.style;
  
    rerender(
      <ThemeProvider>
        <MessageBubble {...defaultProps} isCurrentUser={false} />
      </ThemeProvider>
    );
  
    const otherMessageContainer = getByTestId('message-bubble-container');
    expect(otherMessageContainer).toBeTruthy(); // Asegura que existe despu√©s del rerender
  
    const otherStyle = otherMessageContainer.props.style;
  
    expect(selfStyle).not.toEqual(otherStyle);
  });
  
  

  it('renders OptionsMenu component', () => {
    const { getByTestId } = renderWithTheme(<MessageBubble {...defaultProps} />);
    getByTestId('options-menu');
  });

  it('handles messages without reactions', () => {
    const messageWithoutReactions = { ...mockMessage, reactions: [] };
    const { queryByText } = renderWithTheme(
      <MessageBubble {...defaultProps} message={messageWithoutReactions} />
    );
    expect(queryByText('üëç')).toBeNull();
  });
});
