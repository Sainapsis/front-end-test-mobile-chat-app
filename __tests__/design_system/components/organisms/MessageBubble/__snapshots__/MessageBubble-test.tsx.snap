// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageBubble Component renders correctly with snapshot 1`] = `
[
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessible={true}
    focusable={false}
    onClick={[Function]}
    onLayout={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      [
        {
          "marginVertical": 4,
          "maxWidth": "80%",
        },
        {
          "alignSelf": "flex-end",
        },
      ]
    }
    testID="message-bubble-container"
  >
    <View
      style={
        [
          {
            "borderRadius": 12,
            "boxShadow": "0px 1px 1px rgba(0, 0, 0, 0.1)",
            "elevation": 1,
            "paddingHorizontal": 12,
            "paddingVertical": 8,
          },
          {
            "borderBottomRightRadius": 2,
          },
          {
            "backgroundColor": "#E5E5EA",
          },
        ]
      }
    >
      <Text>
        Hello, this is a test message
      </Text>
      <View
        style={
          {
            "flexDirection": "row",
            "justifyContent": "flex-end",
            "marginTop": 2,
          }
        }
      >
        <Text>
          07:00 PM
        </Text>
      </View>
      <View
        style={
          [
            {
              "backgroundColor": "#FFFFFF",
              "borderRadius": 9999,
              "bottom": -16,
              "elevation": 5,
              "flexDirection": "row",
              "flexWrap": "wrap",
              "gap": 2,
              "padding": 2,
              "position": "absolute",
              "shadowColor": "#212121",
              "shadowOffset": {
                "height": 2,
                "width": 0,
              },
              "shadowOpacity": 0.25,
              "shadowRadius": 3.84,
            },
            {
              "right": 4,
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "backgroundColor": "#F5F5F5",
              "borderRadius": 9999,
              "paddingHorizontal": 2,
              "paddingVertical": 2,
            }
          }
        >
          <Text>
            üëç
          </Text>
        </View>
      </View>
    </View>
  </View>,
  <View
    testID="options-menu"
  />,
]
`;
